### PR TITLE
[WIP] Infection Gamemode (upgraded blob + war ops styled)

### PR DESCRIPTION
**Pull request by Whoneedspacee**
## About The Pull Request

**Credits to @FuryMcFlurry for the gamemodes icons, wouldn't be nearly as cool without him**

Adds a new antagonist, The Infection. When spawned, the station is alerted, and a certain amount of time later (most likely 10 minutes) the core lands a bit off the station. 

There are also beacon generators which generate barriers that prevent the infection from crossing them and will instantly destroy any that do manage to cross it, only taking damage from the infection expanding onto it.

The new infection has many different things, like starting off with at least one evolving spore controlled by a player. These are quite like blobbernauts in blob but they respawn after a certain amount of time after they die, and can select any node to respawn from.

They also are allowed to upgrade and change into different paths in order to counter certain station tactics and avoid situations like emitter spam being the be all end all for winning the game. The infection itself also possesses things like this in the form of being able to upgrade many of their structures, and spend upgrade points gained from destroying the beacons defending the station to gain special abilities or create more evolving spore allies.

A direct reason for this gamemodes existence is preventing early and lame endings to station combats such as nuclear operatives, or situations where the objective can be easily rushed and the round can be over in a short time without many noticing. Which is why the core is immune to all forms of damage or destruction other than from legendary weapons.

Legendary weapons are garnered from killing bosses in events which spawn every 10 minutes starting when the core of the infection lands. They're generally special weapons with special abilities that make it easier to fight the infection or provide support for allies. 

A big thing to note as well is that unlike blob, humans are fully able to move through the infection (with a speed reduction)

## Why It's Good For The Game

This game is generally lacking in a gamemode with a teamwork oriented objective that cannot be rushed. It provides a unique outlet in which you are well aware that everyone around you on the station is your ally, and can work together to your full potential.

There are many things that need to be achieved for a successful round and most will typically involve requiring help of others. If you want to have science past the 10 minute mark you will most likely need to move the machines further back into the station.

If you want power, engineering will likely have to set up generators in order to have power once engineering is taken over. Overall I can't predict how successful a gamemode will be, how fun it will be, or the effort that people are willing to put into it, but I've never seen a gamemode with this much team oriented effort yet, and I think it's a unique opportunity for people to build mechs and create tons of botany guns or other chemicals for people without it being for the goal of destroying something.

Pretty much everyone can try as hard as they want to powergame as much as they can to fight the infection since it has the unique factor of being a clear objective unlike most gamemodes in this game.

**Current Version Video Infection v7**
**||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||**
**||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||**
https://www.youtube.com/watch?v=s1a1o8RxLpw&feature=youtu.be
**||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||**
**||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||||**

## Changelog
:cl: Joe Berry
add: Added The Infection as an antagonist and a gamemode. This is a station defense gamemode against a creature like the blob, except much stronger with upgrades and player controlled minions.
config: Added infection to the rotation with a weight equal to nuclear operatives.
/:cl:

### CODE REVIEW

**Comment by nemvar** [comment]
first

**Comment by Shadowflame909** [comment]
Hell yeah. I can already see tons of ghost role fun.

But I can also see a bunch of people suiciding for said ghost role fun.

**Comment by Whoneedspacee** [comment]
I’m planning on ghost role fun for the station too in terms of centcomm / nanotrasen reinforcements. Not totally sure how I want to implement that just yet, if you have any ideas let me know.

**Comment by habader** [comment]
wew lad

File: code/game/gamemodes/infection/infection.dm [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	<span class='notice'>Crew</span>: Defend long enough until you find a way to destroy the core."
+
+	var/list/pre_spores = list()
+
+	var/commander_datum_type = /datum/antagonist/infection
+
+	var/spore_datum_type = /datum/antagonist/infection/spore
+
+/datum/game_mode/infection/pre_setup()
+	if(!GLOB.infection_spawns.len)
```
Comment (optimumtact) [inline code/game/gamemodes/infection/infection.dm]: why is this a global list?

Comment (Whoneedspacee) [reply code/game/gamemodes/infection/infection.dm]: It’s the spawn positions which are waypoints in the map. It’s used by the antagonist to spawn in the powers file. Similar to what’s done here https://github.com/tgstation/tgstation/blob/25f18ff5232f6db4ec6806535eed4a98e11e984b/code/modules/antagonists/nukeop/nukeop.dm#L97

**Comment by Tlaltecuhtli** [comment]
blob and war ops? just what i needed

**Comment by MrFluffster** [comment]
Please use the enter key, it really helps with reading when the text is broken up into manageable chunks.
Nobody likes blocks of text.

**Comment by knacker48** [comment]
If this comes in will it completely replace blob as it is now?

**Comment by wesoda25** [comment]
I don’t really have an opinion on most of this, but I think it’s important to note the waiting period is the reason a lot of people hate nuke ops and had it set to such a low chance. A boring grace period followed by an antag which seems even less flexible than nuke ops doesn’t seem like it’ll age well.

**Comment by davethwave** [comment]
Might be good to replace the turret sprites, think something like the anti air turret things from the zerg could be good inspiration to go after with how this gamemode works.

**Comment by Whoneedspacee** [comment]
pretty much everything is placeholder  in terms of icons

**Comment by optimumtact** [comment]
Please reduce your changelog to the first line, don't try to stuff a gamemode tutorial in it, that's what the wiki is for.

**Comment by Whoneedspacee** [comment]
got it

**Comment by OverlyGeneric** [comment]
This is like the Purge but for powergamers to go all out just to get out their system.

**Comment by Whoneedspacee** [comment]
> I don’t really have an opinion on most of this, but I think it’s important to note the waiting period is the reason a lot of people hate nuke ops and had it set to such a low chance. A boring grace period followed by an antag which seems even less flexible than nuke ops doesn’t seem like it’ll age well.

I'm very much considering lowering the waiting period to an extremely low amount such as 2 minutes with the excuse that they were unable to track it until it entered the solar system, and just making the infection much weaker off the bat. It would provide much more of a tug of war and slow build up for both sides. Especially since the defensive beacons along with weaker enemies makes it harder for the infection to advance.

I also plan on making it easier to fight the infection by giving gravity while you're inside it as well as not requiring any atmospheric suits while you're inside it as well, probably something the infection could change with an evolution for some added variability.

By very much considering I mean all future design is going towards this unless I find some major problem with it.

**Comment by Whoneedspacee** [comment]
damn you blob refactor fucking my shit up

File: code/datums/components/nobeaconcrossing.dm [hunk 2]
```diff
@@ -0,0 +1,43 @@
+/datum/component/no_beacon_crossing
+	var/atom/parentatom
```
Comment (ninjanomnom) [inline code/datums/components/nobeaconcrossing.dm]: Why do people keep trying to do this, stop it. Typecast when you need it don't store your parent again

File: code/datums/components/nobeaconcrossing.dm [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+/datum/component/no_beacon_crossing/Initialize()
+	parentatom = parent
```
Comment (ninjanomnom) [inline code/datums/components/nobeaconcrossing.dm]: cease

File: code/datums/components/chasm.dm [hunk 4]
```diff
@@ -51,8 +51,13 @@
 	. = 0
 	if (is_safe())
 		return FALSE
-
 	var/atom/parent = src.parent
+	if(locate(/obj/structure/infection) in parent.contents)
+		// no thanks that would be stupid to fall in over that or after destroying it
+		var/turf/T = get_turf(parent)
+		T.ChangeTurf(/turf/open/floor/holofloor/asteroid)
```
Comment (ninjanomnom) [inline code/datums/components/chasm.dm]: Please don't put snowflake code in components like this, why is this here?

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

File: code/modules/antagonists/infection/infection/theinfection.dm [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+	if(overmind)
+		add_atom_colour(overmind.infection_color, FIXED_COLOUR_PRIORITY)
+	else
+		remove_atom_colour(FIXED_COLOUR_PRIORITY)
+
+/obj/structure/infection/process()
+	Life()
+
+/obj/structure/infection/proc/Life()
+	SEND_SIGNAL(src, COMSIG_INFECTION_LIFE_TICK)
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/theinfection.dm]: Absolutely not allowed, no signals in processing procs.

File: code/modules/antagonists/infection/infection/infections/upgrades.dm [hunk 6]
```diff
@@ -0,0 +1,398 @@
+/datum/component/infection/upgrade
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/infections/upgrades.dm]: components go in the component folder

File: code/modules/antagonists/infection/infection/infections/upgrades.dm [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+	var/times = 1 // times the upgrade can be bought
+	var/bought = 0 // how many times the upgrade has been bought
+
+/datum/component/infection/upgrade/Initialize()
+	. = ..()
+	if(bought)
+		register_signals()
+
+/datum/component/infection/upgrade/proc/register_signals()
+	RegisterSignal(parent, COMSIG_INFECTION_LIFE_TICK, .proc/life_tick)
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/infections/upgrades.dm]: If something needs to process it should not piggyback off another thing processing. Add it to a subsystem on its own if it needs to process.

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by ninjanomnom** [comment]
Contact me in discord/irc/whatever, there's a lot I need to talk to you about in regards to your use of components here

File: code/game/gamemodes/infection/infection.dm [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+		var/datum/mind/spore_mind = pre_spores[i]
+		spore_mind.add_antag_datum(spore_datum_type)
+	return ..()
+
+/datum/game_mode/infection/check_finished()
+	var/mob/camera/commander/C = GLOB.infection_commander
+	// end if commander is dead, or if beacons are destroyed (handled in antagonist win check)
+	if(C)
+		return FALSE
+	return TRUE
```
Comment (ninjanomnom) [inline code/game/gamemodes/infection/infection.dm]: `return !isnull(GLOB.infection_commander)`

File: code/modules/antagonists/infection/beacon.dm [hunk 9]
```diff
@@ -0,0 +1,158 @@
...
+
+GLOBAL_LIST_EMPTY(infection_beacons)
+GLOBAL_LIST_EMPTY(beacon_spawns)
+
+/obj/effect/landmark/beacon_start
+	name = "beaconstart"
+	icon_state = "beacon_start"
+
+/obj/effect/landmark/beacon_start/Initialize(mapload)
+	..()
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/beacon.dm]:
```suggestion
	. = ..()
```

File: code/modules/antagonists/infection/beacon.dm [hunk 10]
```diff
@@ -0,0 +1,158 @@
...
+	density = TRUE
+	anchored = TRUE
+	layer = FLY_LAYER
+	CanAtmosPass = ATMOS_PASS_PROC
+	max_integrity = 1000
+	resistance_flags = INDESTRUCTIBLE
+	var/list/walls = list()
+
+/obj/structure/beacon_generator/Initialize(mapload)
+	START_PROCESSING(SSobj,src)
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/beacon.dm]: Why do this before the parent call?

File: code/modules/antagonists/infection/beacon.dm [hunk 11]
```diff
@@ -0,0 +1,158 @@
...
+		qdel(src)
+		for(var/i = 1 to 5)
+			for(var/atom/A in urange(i, explodeloc) - urange(i - 1, explodeloc))
+				A.ex_act(EXPLODE_LIGHT)
+				if(istype(A, /obj/structure/infection))
+					var/obj/structure/infection/INF = A
+					INF.take_damage(1000, BRUTE, "bomb", 0)
+			sleep(4)
+	else
+		playsound(src.loc, 'sound/effects/empulse.ogg', 300, 1, 10, pressure_affected = FALSE)
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/beacon.dm]:
```
if(obj_integrity > 0)
  playsound(...
  return
otherstuff...
```

File: code/modules/antagonists/infection/infection/commander.dm [hunk 12]
```diff
@@ -0,0 +1,238 @@
...
+			stat(null, "Core Health: [infection_core.obj_integrity]")
+			stat(null, "Power Stored: [infection_points]/[max_infection_points]")
+			stat(null, "Upgrade Points: [upgrade_points]")
+			stat(null, "Beacons Remaining: [GLOB.infection_beacons.len]")
+		if(!placed)
+			stat(null, "Time Before Automatic Placement: [max(round((autoplace_time - world.time)*0.1, 0.1), 0)]")
+
+/mob/camera/commander/Move(NewLoc, Dir = 0)
+	if(meteor)
+		return 0
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/commander.dm]: Use FALSE and TRUE always for bools
```suggestion
		return FALSE
```

File: code/modules/antagonists/infection/infection/commander.dm [hunk 13]
```diff
@@ -0,0 +1,238 @@
...
+	forceMove(NewLoc)
+	return 1
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/commander.dm]:
```suggestion
	return TRUE
```

File: code/modules/antagonists/infection/infection/infections/core.dm [hunk 14]
```diff
@@ -0,0 +1,168 @@
...
+		update_icon()
+		for(var/mob/living/simple_animal/hostile/infection/infectionspore/sentient/S in overmind.contents)
+			S.forceMove(get_turf(src))
+	point_rate = new_rate
+	addtimer(CALLBACK(src, .proc/generate_announcement), 40)
+	SSevents.frequency_lower = DOOM_CLOCK_EVENT_DELAY
+	SSevents.frequency_upper = DOOM_CLOCK_EVENT_DELAY
+	SSevents.toggleInfectionmode()
+	SSevents.reschedule()
+	. = ..()
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/infections/core.dm]: Please have the parent call as early as possible in the proc

File: code/modules/antagonists/infection/infection/infections/core.dm [hunk 15]
```diff
@@ -0,0 +1,168 @@
...
+	cut_overlays()
+	color = null
+	var/mutable_appearance/infection_overlay = mutable_appearance('icons/mob/blob.dmi', "blob")
+	if(overmind)
+		infection_overlay.color = overmind.infection_color
+	add_overlay(infection_overlay)
+	add_overlay(mutable_appearance('icons/mob/blob.dmi', "blob_core_overlay"))
+
+/obj/structure/infection/core/Destroy()
+	. = ..()
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/infections/core.dm]: ...except destroy which needs to be last

File: code/modules/antagonists/infection/infection/infections/core.dm [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+	STOP_PROCESSING(SSobj, src)
+	GLOB.poi_list -= src
+	SSevents.toggleInfectionmode()
+
+/obj/structure/infection/core/proc/deathExplosion()
+	var/explodeloc = src.loc
+	for(var/i = 1 to 9)
+		for(var/atom/A in urange(i, explodeloc) - urange(i - 1, explodeloc))
+			var/power = CEILING(i/3, 1)
+			A.ex_act(power)
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/infections/core.dm]: Why are you handrolling your own explosion?

File: code/modules/antagonists/infection/infection/infections/core.dm [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+			if(istype(A, /obj/structure/infection))
+				var/obj/structure/infection/INF = A
+				INF.take_damage(600 / power, BRUTE, "bomb", 0)
+		sleep(4)
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/infections/core.dm]: Switch this to be multiple timers

**Review by ninjanomnom — CHANGES_REQUESTED** [review]
Still more but I've been interrupted

File: code/game/gamemodes/infection/infection.dm [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+/datum/game_mode/infection/pre_setup()
+	if(!GLOB.infection_spawns.len)
+		setup_error = "No infection core spawnpoints found"
+		return FALSE
+	if(!GLOB.beacon_spawns.len)
+		setup_error = "No infection beacon spawnpoints found"
+		return FALSE
+	var/n_spores = min(CEILING(num_players() / 10, 1), antag_candidates.len)
+	if(n_spores >= required_enemies || GLOB.Debug2)
+		for(var/i = 0, i < n_spores, ++i)
```
[the remark this answers is not in the document]
Comment (ninjanomnom) [reply code/game/gamemodes/infection/infection.dm]: Or just `1 to n_spores` if you change the other stuff

**Comment by FuryMcFlurry** [comment]
image

Blobs are dum, crystals are cool.

**Comment by Whoneedspacee** [comment]
thank fury for the new icons he's doing

**Comment by zxaber** [comment]
>The shadow realm has come to claim the station

Looks good, for real

File: code/datums/components/nobeaconcrossing.dm [hunk 19]
```diff
@@ -0,0 +1,40 @@
+/datum/component/no_beacon_crossing/Initialize()
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/check_passed)
```
Comment (ninjanomnom) [inline code/datums/components/nobeaconcrossing.dm]: `return COMPONENT_INCOMPATIBLE` when parent isnt a movable

File: code/datums/components/nobeaconcrossing.dm [hunk 20]
```diff
@@ -0,0 +1,40 @@
...
+		should_die = TRUE
+	if(facingdir == SOUTH && edge.y >= parentatom.y)
+		should_die = TRUE
+	if(facingdir == EAST && edge.x >= parentatom.x)
+		should_die = TRUE
+	if(facingdir == WEST && edge.x <= parentatom.x)
+		should_die = TRUE
+	if(isobj(parentatom.loc))
+		should_die = FALSE // don't kill them if they're in a locker, or something is holding them
+	if(should_die)
```
Comment (ninjanomnom) [inline code/datums/components/nobeaconcrossing.dm]: Make this a seperate proc like kill_parent() or something

File: code/modules/antagonists/infection/beacon.dm [hunk 21]
```diff
@@ -0,0 +1,158 @@
...
+		return
+	playsound(src.loc, 'sound/magic/repulse.ogg', 300, 1, 10, pressure_affected = FALSE)
+	var/mob/camera/commander/OM = GLOB.infection_commander
+	OM.playsound_local(OM, 'sound/magic/repulse.ogg', 300, 1)
+	var/explodeloc = src.loc
+	qdel(src)
+	for(var/i = 1 to 5)
+		for(var/atom/A in urange(i, explodeloc) - urange(i - 1, explodeloc))
+			A.ex_act(EXPLODE_LIGHT)
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/beacon.dm]: Another case of "why are you handrolling an explosion"

File: code/modules/antagonists/infection/beacon.dm [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+			if(istype(A, /obj/structure/infection))
+				var/obj/structure/infection/INF = A
+				INF.take_damage(1000, BRUTE, "bomb", 0)
+		sleep(4)
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/beacon.dm]: 5 timers 4 ticks apart each instead of this sleep

File: code/modules/antagonists/infection/infection.dm [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+	name = "Infection"
+	roundend_category = "infections"
+	antagpanel_category = "Infection"
+	job_rank = ROLE_INFECTION
+
+	var/starting_points_human_infection = 60
+	var/point_rate_human_infection = 2
+
+/datum/antagonist/infection/roundend_report()
+	var/basic_report = ..()
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection.dm]: You're allowed to change `.` you know. `. = ..()` works just fine here

File: code/modules/antagonists/infection/infection.dm [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+	owner.current.playsound_local(get_turf(owner.current), 'sound/effects/blobattack.ogg',100,0)
+	owner.current.playsound_local(get_turf(owner.current), 'sound/effects/attackblob.ogg',100,0)
+	to_chat(owner,"<span class='userdanger'>You feel radiant.</span>")
+
+/datum/antagonist/infection/on_gain()
+	create_objectives()
+	var/turf/start = pick(GLOB.infection_spawns)
+	var/mob/newmob = create_mob_type(start)
+	owner.transfer_to(newmob, TRUE)
+	. = ..()
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection.dm]: If this needs to be the last line do `return ..()` instead

File: code/modules/antagonists/infection/infection/commander.dm [hunk 25]
```diff
@@ -0,0 +1,238 @@
...
+	last_attack = world.time
+	if(infection_core)
+		infection_core.update_icon()
+	SSshuttle.registerHostileEnvironment(src)
+	addtimer(CALLBACK(src, .proc/generate_announcement), place_beacons_delay / 2)
+	addtimer(CALLBACK(src, .proc/place_beacons), place_beacons_delay)
+	for(var/type_action in default_actions)
+		var/datum/action/innate/infection/add_action = new type_action()
+		add_action.Grant(src)
+	.= ..()
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/commander.dm]: Why is this so late in initialize? None of this looks like it needs to be before parent inits?

File: code/modules/antagonists/infection/infection/commander.dm [hunk 26]
```diff
@@ -0,0 +1,238 @@
...
+	STOP_PROCESSING(SSobj, src)
+
+	SSshuttle.clearHostileEnvironment(src)
+
+	addtimer(CALLBACK(src, .proc/defeated_announcement), 80)
+
+	return ..()
+
+/mob/camera/commander/Login()
+	..()
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/commander.dm]: default to `. = ..()`
```suggestion
	. = ..()
```

File: code/modules/antagonists/infection/infection/commander.dm [hunk 27]
```diff
@@ -0,0 +1,238 @@
...
+	to_chat(src, "<span class='notice'>You are the infection!</span>")
+	infection_help()
+	update_health_hud()
+	add_points(0)
+
+/mob/camera/commander/examine(mob/user)
+	..()
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/commander.dm]:
```suggestion
	. = ..()
```

File: code/modules/antagonists/infection/infection/infections/core.dm [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+
+/obj/structure/infection/core/Initialize(mapload, client/new_overmind = null, new_rate = 2, placed = 0)
+	if(GLOB.infection_core)
+		return INITIALIZE_HINT_QDEL // just making sure admins can't break everything
+	GLOB.infection_core = src
+	GLOB.poi_list |= src
+	update_icon() //so it atleast appears
+	if(!placed && !overmind)
+		return INITIALIZE_HINT_QDEL
+	. = ..()
```
Comment (ninjanomnom) [inline code/modules/antagonists/infection/infection/infections/core.dm]: Again, why is this so late in initialize?

**Review by ninjanomnom — CHANGES_REQUESTED** [review]

**Comment by ArcaneMusic** [comment]
Holy shit the sprites look really good, like damn. Some of the textures for walls and floors could maybe look a bit more jagged, perhaps, but I'm still really pleased with the theming for this gamemode being realized.

**Comment by Whoneedspacee** [comment]
https://www.youtube.com/watch?v=s1a1o8RxLpw&feature=youtu.be

Another update video, been a while eh?

File: code/datums/components/nobeaconcrossing.dm (hunk 20 above)
Comment (ninjanomnom) [reply code/datums/components/nobeaconcrossing.dm]: You've made the proc, now you dont need to set should_die and can just directly call `kill_parent()`

**Comment by SpaceManiac** [comment]
Closing as Stale for being Work In Progress for over a week with no updates. If you want to continue work, reply here or ping a maintainer.
